shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs

```diff
@@ -103,68 +70,281 @@ async fn topology_task_process(
+                    metadata,
+                }) => {
+                    let (
+                        broadcast_address_index,
+                        rack_index,
+                        data_center_index,
+                        tokens_index,
+                        broadcast_port_index,
+                    ) = {
+                        let mut broadcast_address_index: Option<usize> = None;
+                        let mut rack_index: Option<usize> = None;
+                        let mut data_center_index: Option<usize> = None;
+                        let mut tokens_index: Option<usize> = None;
+                        let mut broadcast_port_index: Option<usize> = None;
+
+                        metadata
+                            .col_specs
+                            .iter()
+                            .enumerate()
+                            .for_each(|(i, col_spec)| match col_spec.name.as_str() {
+                                "broadcast_address" => broadcast_address_index = Some(i),
+                                "rack" => rack_index = Some(i),
+                                "data_center" => data_center_index = Some(i),
+                                "tokens" => tokens_index = Some(i),
+                                "rpc_port" => broadcast_port_index = Some(i),
+                                _ => {}
+                            });
+
+                        (
+                            broadcast_address_index
+                                .context("broadcast_address missing from columns")?,
+                            rack_index.context("rack missing from columns")?,
+                            data_center_index.context("data_center missing from columns")?,
+                            tokens_index.context("tokens missing from columns")?,
+                            broadcast_port_index,
+                        )
                     };
 
-                    Ok(CassandraNode {
-                        address: SocketAddr::new(ip, 9042),
-                        rack,
-                        _tokens: tokens,
-                        outbound: None,
+                    rows.iter_mut()
+                        .filter(|row| {
+                            if let Some(MessageValue::Varchar(data_center)) =
+                                row.get(data_center_index)
+                            {
+                                data_center == config_data_center
+                            } else {
+                                false
+                            }
+                        })
+                        .map(|row| {
+                            let tokens =
+                                if let Some(MessageValue::List(list)) = row.get(tokens_index) {
+                                    list.iter()
+                                        .map::<Result<String>, _>(|x| match x {
+                                            MessageValue::Varchar(a) => Ok(a.clone()),
```

Comment:
You can avoid cloning all of these strings like this:
![image](https://user-images.githubusercontent.com/5120858/188768051-e22569ca-3edc-49c3-9ffb-e9a5283f09ee.png)